features/netsocket/nsapi_dns.cpp

```diff
@@ -305,3 +519,337 @@ nsapi_error_t nsapi_dns_query(NetworkStack *stack, const char *host,
+            query = dns_query_queue[i];
+            break;
+        }
+    }
+
+    if (!query) {
+        nsapi_dns_query_async_resp(query, NSAPI_ERROR_NO_MEMORY, NULL);
+        return;
+    }
+
+    if (query->retries) {
+        query->retries--;
+    } else {
+        query->dns_server++;
+        query->retries = 1;
+    }
+
+    query->dns_message_id = dns_message_id++;
+
+    // create network packet
+    uint8_t *packet = (uint8_t *)malloc(DNS_BUFFER_SIZE);
+    if (!packet) {
+        nsapi_dns_query_async_resp(query, NSAPI_ERROR_NO_MEMORY, NULL);
+        return;
+    }
+
+    // send the question
+    int len = dns_append_question(packet, query->dns_message_id, query->host, query->version);
+
+    while (true) {
+        SocketAddress dns_addr;
+        nsapi_size_or_error_t err = nsapi_dns_get_server_addr(query->stack, &(query->dns_server), &dns_addr);
+        if (err != NSAPI_ERROR_OK) {
+            nsapi_dns_query_async_resp(query, NSAPI_ERROR_DNS_FAILURE, NULL);
+            free(packet);
+            return;
+        }
+
+        err = query->socket->sendto(dns_addr, packet, len);
+
+        if (err < 0) {
+            query->dns_server++;
+        } else {
+            break;
+        }
+    }
+
+    free(packet);
+
+    if (nsapi_dns_call_in(query->stack, DNS_TIMEOUT,
+        mbed::callback(nsapi_dns_query_async_send, reinterpret_cast<void *>(unique_id))) != NSAPI_ERROR_OK) {
+        nsapi_dns_query_async_resp(query, NSAPI_ERROR_NO_MEMORY, NULL);
+    }
+}
+
+static void nsapi_dns_query_async_socket_callback(NetworkStack *stack)
+{
+    nsapi_dns_call(stack, mbed::callback(nsapi_dns_query_async_socket_callback_handle, stack));
+}
+
+static void nsapi_dns_query_async_socket_callback_handle(NetworkStack *stack)
+{
+    UDPSocket *socket = NULL;
+
+    for (int i = 0; i < DNS_QUERY_QUEUE_SIZE; i++) {
+        if (dns_query_queue[i] && dns_query_queue[i]->stack == stack) {
+            socket = dns_query_queue[i]->socket;
+            break;
+        }
+    }
+
+    if (socket) {
+        // create network packet
+        uint8_t *packet = (uint8_t *)malloc(DNS_BUFFER_SIZE);
+        if (!packet) {
+            return;
+        }
+
+        // recv the response
+        nsapi_size_or_error_t size = socket->recvfrom(NULL, packet, DNS_BUFFER_SIZE);
+
+        if (size < DNS_RESPONSE_MIN_SIZE) {
+            free(packet);
+            return;
+        }
+
+        // gets id from response to associate with correct query
+        uint16_t id = (*packet << 8) | *(packet + 1);
+
+        DNS_QUERY *query = NULL;
+
+        for (int i = 0; i < DNS_QUERY_QUEUE_SIZE; i++) {
+            if (dns_query_queue[i] && dns_query_queue[i]->dns_message_id == id) {
+                query = dns_query_queue[i];
+                break;
+            }
+        }
+
+        if (!query) {
+            free(packet);
+            return;
+        }
+
+        nsapi_addr_t *addrs = new nsapi_addr_t[query->addr_count];
```

Comment:
Mixing C/C++ allocator types is a bit odd. Not going to rule it out, but these C++ ones should probably be `nothrow` with null checks.